src/mgmt/srv_query.c

```diff
@@ -0,0 +1,192 @@
+/**
+ * (C) Copyright 2016-2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+/*
+ * ds_mgmt: Storage Query Methods
+ */
+#define D_LOGFAC	DD_FAC(mgmt)
+
+#include <daos_srv/bio.h>
+#include <daos_srv/smd.h>
+
+#include "srv_internal.h"
+
+static void
+bio_health_query(void *arg)
+{
+	struct mgmt_bio_health	*mbh = arg;
+	struct dss_module_info	*info = dss_get_module_info();
+	struct bio_xs_context	*bxc;
+
+	D_ASSERT(info != NULL);
+	D_DEBUG(DB_MGMT, "BIO health stats query on xs:%d, tgt:%d\n",
+		info->dmi_xs_id, info->dmi_tgt_id);
+
+	bxc = info->dmi_nvme_ctxt;
+	if (bxc == NULL) {
+		D_ERROR("BIO NVMe context not initialized for xs:%d, tgt:%d\n",
+			info->dmi_xs_id, info->dmi_tgt_id);
+		goto out;
+	}
+
+	mbh->mb_dev_state = bio_get_dev_state(bxc);
+	if (mbh->mb_dev_state == NULL) {
+		D_ERROR("Error getting BIO device state\n");
+		goto out;
+	}
+
+out:
+	smd_free_dev_info(mbh->mb_dev_info);
+}
+
+int
+ds_mgmt_bio_health_query(struct mgmt_bio_health *mbh, uuid_t dev_uuid,
+			char *tgt)
+{
+	ABT_thread		 thread;
+	int			 tgt_id;
+	int			 rc = 0;
+
+	if (uuid_is_null(dev_uuid) && strlen(tgt) == 0) {
+		/* Either dev uuid or tgt id needs to be specified for query */
+		D_ERROR("Neither dev_uuid or tgt_id specified for BIO query\n");
+		return -DER_INVAL;
+	}
+
+	/*
+	 * Query per-server metadata (SMD) to get either target ID(s) for given
+	 * device or alternatively the device mapped to a given target.
+	 */
+	if (!uuid_is_null(dev_uuid)) {
+		rc = smd_dev_get_by_id(dev_uuid, &mbh->mb_dev_info);
```

Comment:
The mbh->mb_dev_info is leaked.

Reply:
will add goto out to free dev_info upon error and successful completion of funtion